### ISSUE TITLE
Consolidate code of core component into one repo

### ISSUE THREAD

**Issue by qiujian16** [open-cluster-management-io/ocm#128]
We currently have different component repos:
- registration
- work
- placement
- registration-operator
- clusteradm
- addon-framework

It is hard to do e2e tests with many components and hard to do dependency check. I propose,we consolidate the code in these repos to one single repo:

- code in the work/registration/placement/operator/clusteradm component is merged into https://github.com/open-cluster-management-io/OCM
- split addon-framework to pure library code and addon-manager code. The addon-manager code is also put into https://github.com/open-cluster-management-io/OCM
- archive the the independent repo

After that we will have
- https://github.com/open-cluster-management-io/api
- https://github.com/open-cluster-management-io/OCM
- https://github.com/open-cluster-management-io/addon-framework (just library, remove unnecessary deps)

With this consolidation, we will have the benefit that
1. we do not need webhook per component, we can have one single webhook server
2. we can build hub and agent as one single binary
3. easier e2e testing
4. better understanding and dev experience for the contributor

note from @deads2k we should validate gates for PR that we do not have cross deps among different components.

**Comment by clyang82** [comment]
Can we rename `https://github.com/open-cluster-management-io/OCM` to `https://github.com/open-cluster-management-io/ocm`? It seems the lower-case repo name is a common practices.

**Comment by xuezhaojun** [comment]
https://github.com/xuezhaojun/ocm-core: An example merged repo, keeping the same commit history.

**Comment by xuezhaojun** [comment]
> I'm on board with consolidating the repos, but I'm not sure having `clusteradm` in the same repo with the controllers makes sense. It seems like the CLI should have its own repo since it's not a controller/API?

But it seems including CLI in the main repo is a common pattern, is there particular advantages we can gain from not include it?

**Comment by dhaiducek** [comment]
I wasn't aware including the CLI in the main repo was a common pattern (`kubectl` was the only one I checked). I was thinking having it separate might make it more discoverable and would keep the `ocm` repo focused on controllers.

**Comment by xuezhaojun** [comment]
I checked a few repos like [argo-cd](https://github.com/argoproj/argo-cd/blob/5bdd42ceed96de92667262cf066aafffa2891be6/util/cli/cli.go), [istio](https://github.com/istio/istio/tree/master/istioctl) and one similar usage repo [karmada](https://github.com/karmada-io/karmada/blob/master/cmd/karmadactl/karmadactl.go). 

It may not be a common pattern, but if we separate it, we have to maintain 2 e2e tests.  If we keep it in `ocm`, any changes would trigger test cases in the controller and CLI, and we gain confidence in the final user experience.

**Comment by mprahl** [comment]
> I checked a few repos like [argo-cd](https://github.com/argoproj/argo-cd/blob/5bdd42ceed96de92667262cf066aafffa2891be6/util/cli/cli.go), [istio](https://github.com/istio/istio/tree/master/istioctl) and one similar usage repo [karmada](https://github.com/karmada-io/karmada/blob/master/cmd/karmadactl/karmadactl.go).
> 
> It may not be a common pattern, but if we separate it, we have to maintain 2 e2e tests. If we keep it in `ocm`, any changes would trigger test cases in the controller and CLI, and we gain confidence in the final user experience.

Note that you can reuse GitHub workflows from another repo so you could run the same tests in both repos without duplicating them.

**Comment by xuezhaojun** [comment]
Hello, everyone, since we are going to reach release 0.11.0, now it's a good time to discuss the next executable steps of code consolidation. I made a draft plan, please take a review when you have time, Thanks!

- [ ] Release `0.11.0` @qiujian16 
- [ ] Rename `OCM` to `ocm`: https://github.com/open-cluster-management-io/OCM/issues/139
- [ ] Merge the Pull Request that catches up latest commits: https://github.com/open-cluster-management-io/OCM/pull/137
- [ ] Merge the branch `code-consolidation` to `main`.
  *  After this, devs are able to develop new features on the `OCM` repo.  
- [ ] Disable origin repos' GitHub actions on the main branch.
- [ ] Merge the Pull Request that adds GitHub actions: https://github.com/open-cluster-management-io/OCM/pull/132
  * After this, devs are able to push new PRs into OCM.
  * The `pre+post` and `release` should separate into 2 PRs. `release` should test locally first.
- [ ] Announce in channel and mail group. 
- [ ] Update Readme in the `ocm` repo and origin repos.
  * Provide a `Dev Guidence` in the main branch in the`ocm`.
  * Add a note mentioning dev works moved to the `ocm` repo in origin repos.
- [ ] Update website references.

**Comment by xuezhaojun** [comment]
We have a new PR to relocate code out of staging: https://github.com/open-cluster-management-io/ocm/pull/146
The code needs further merging in the future to reduce redundancy.
 
## Change Notes

### For all components:
We left the e2e-tests and some helper scripts still under the `staging/` folder. If you want to run e2e, for example, placement, try a command like: `make -C ./staging/src/open-cluster-management.io/placement test-e2e` to specify the Makefile.

### For `registration-operator`:
We have `manifests/`, `deploy`, and `hack/` of the origin registration-operator repo copied in `ocm`. The content is unchanged.
 
### For `placement`, `registration`, and `work`:
 * The integration-test used to depend on CRDs files under `/deploy`, now it's depending on CRDs in vendor directly. For example, in `placement`:
    * Before: https://github.com/open-cluster-management-io/placement/blob/2572e0a55c2d4635b372fbd7dd7382e35983cd56/test/integration/suite_test.go#L39 
    * After:  https://github.com/open-cluster-management-io/ocm/blob/ad38b9465f64fb107b67f7b24f2dfe0e1fda3322/test/placement/integration/suite_test.go#L46.
 * We have a unified integration-test makefile now, the `make test-integration` is run tests of each component one by one: https://github.com/open-cluster-management-io/ocm/blob/ad38b9465f64fb107b67f7b24f2dfe0e1fda3322/test/integration-test.mk#L49. 
 * Use `make build` can build binaries for all 4 components, Use `make test` can run unit-test for all 4 components.
 
### For `placement`: @haoqing0110 
 * The benchmark depends on pkg in integration, so we move `util` pkg up a level. 
    * Before: https://github.com/open-cluster-management-io/placement/tree/main/test/integration/util
    * After: https://github.com/open-cluster-management-io/ocm/tree/main/test/placement/util
 * The [benchmark CRDs is depending on vendor files](https://github.com/open-cluster-management-io/ocm/blob/ad38b9465f64fb107b67f7b24f2dfe0e1fda3322/test/placement/benchmark/benchmark_test.go#L110) as well. You may need to run in the root directory of ocm.
 
### Unsolved or affected:
* Release(including OLM release) @zhiweiyin318 
*

**Comment by qiujian16** [comment]
@xuezhaojun  the progress seems good. I think we should have a discussion and start creating some small tasks. e.g.
- refactor e2e out of staging
- further consolidating the integration
- refactor cmd/dependencymagnet/version code
